File: dfs_bfs_backtracking/cpp/b14889.cpp
#include <iostream>
#include <vector>
int players[20][20];
int N;
int min_gap = 10000;


int get_status(std::vector<int> team) {
	int team_status = 0;
	for (int i = 0; i < team.size(); ++i) {
		for (int j = 0; j < team.size(); ++j) {
			team_status += players[team[i]][team[j]];
		}
	}
	return team_status;
}


void backtracking(std::vector<bool>& team_status, const int& pos, const int& cnt) {
	if (cnt == N / 2) {
		std::vector<int> team1;
		std::vector<int> team2;
		for (int i = 0; i < team_status.size(); ++i) {
			if (team_status[i]) { team1.push_back(i); }
			else { team2.push_back(i); }
		}
		int team1_status = get_status(team1);
		int team2_status = get_status(team2);

		int temp_gap = team1_status > team2_status ? (team1_status - team2_status) : (team2_status - team1_status);
		if (temp_gap < min_gap) {
			min_gap = temp_gap;
		}
		return;
	}

	if ((N/2) < pos - cnt) {
		return;
	}

	team_status[pos] = 1;
	backtracking(team_status, pos + 1, cnt + 1);
	team_status[pos] = 0;
	backtracking(team_status, pos + 1, cnt);
	return;
}

int main() {
	std::cin >> N;
	for (int i = 0; i < N; ++i) {
		for (int j = 0; j < N; ++j) {
			std::cin >> players[i][j];
		}
	}

	std::vector<bool> team_status;
	team_status.resize(N);
	backtracking(team_status, 0, 0);
	
	std::cout << min_gap << std::endl;
}

File: dfs_bfs_backtracking/python/b14889.py
N = int(input())
team_stat = [list(map(int, input().split())) for _ in range(N)]
min_num = 100000
team = [0] * N

def get_status(team_in):
    status = 0
    for p1 in team_in:
        for p2 in team_in:
            status += team_stat[p1][p2]
    return status

def backtracking(team, pos, cnt):
    global min_num
    if cnt == N / 2:
        team1 = []
        team2 = []
        for i in range(N):
            if team[i]:
                team1.append(i)
            else:
                team2.append(i)
        team1_status = get_status(team1)
        team2_status = get_status(team2)
        result = team1_status - team2_status if team1_status > team2_status else team2_status - team1_status
        if result < min_num:
            min_num = result
        return
    
    if N / 2 < pos - cnt:
        return

    team[pos] = 1
    backtracking(team, pos + 1, cnt + 1)
    team[pos] = 0
    backtracking(team, pos + 1, cnt)

backtracking(team, 0, 0)
print(min_num)